### ISSUE TITLE
scope3d~

### ISSUE THREAD

**Issue by porres** [porres/pd-else#2050]
reboot of https://github.com/porres/pd-else/issues/2041
- rename 'delay' to 'rate'... 
- I still don't get 'perspective'... 
- add 'range' message to set display range (could probably replace 'zoom')
- the dimensions, whatever they are, should match the range, so you can stretch it like it is with [oscope~]
- escape dollar signs for builtin receive name

**Comment by ben-wes** [comment]
`rate` is fixed.

don't really know how to better explain `perspective`. but here's a possibly better generated explanation: "Adjusting the perspective setting is like changing the 'depth feel' of the scene. Increasing it makes objects appear more dramatically different in size based on their distance, enhancing the 3D effect. Decreasing it makes the scene look flatter, as if you're moving towards a more 2D appearance."

the intended `range` behaviour is not yet completely clear to me: would it be a stretched scene by default then if you choose a dimension where width != height? (meaning: the "range" would match the dimensions by default?)

**Comment by porres** [comment]
range would actually be similar to zoom I guess, but we'd be able to set range for the 3 dimensions

**Comment by ben-wes** [comment]
should the range then restrict the space for the visuals? i'm not sure i understand ... can't really imagine how this should behave. for 2d, it's obvious - but for a 3d view that you can rotate, i don't see the obvious behaviour. especially for the z-range i wouldn't know what to do.

**Comment by porres** [comment]
nlines and nsamples dont really work like in scope~

**Comment by porres** [comment]
> i wouldn't know what to do.

not sure either

**Comment by ben-wes** [comment]
> nlines and nsamples dont really work like in scope~

yes. i compared a bit more as well now. nlines is the same, i assume. but nsamples is not really clear to me. and it also looks like the drawing behaviour of scope~ is different in general. without really checking the code, it looks like it's not working with a ring buffer, but instead collects all samples based on the nlines/nsamples for every frame. that's why it's updating so slowly with higher settings like `nsamples 256, nlines 256` (it obviously waits for 256*256 = 65536 samples then and updates only every 65536/samplerate seconds). i modestly prefer scope3d~'s behaviour there, haha.

btw, i just noticed that scope~ with these high settings is quite heavily slowing down the gui responsiveness, too ... but the cpu meter is not reflecting this.

**Comment by ben-wes** [comment]
or maybe for a better side by side comparison, these are exaggerated values - better use `nsamples 16, nlines 256` or so. :)

**Comment by porres** [comment]
- better use nsamples 16, nlines 256 or so for what?

**Comment by porres** [comment]
Scope freezes gui and is very bad, not sure how to improve it

**Comment by ben-wes** [comment]
> * better use nsamples 16, nlines 256 or so for what?

for comparing behaviors of the two

**Comment by porres** [comment]
more like '8 256', btw, list input should set both values in scope3d~ as well

**Comment by porres** [comment]
but default in scope3d~ is 'nsamples 8, nlines 512', but setting it also with  '8 256' it doesn't look different... anyway, I'm testing things now and it seems both settings are actually in fact the same... both do the same and look the same now... wtf?

**Comment by porres** [comment]
[compare.pd.zip](https://github.com/porres/pd-else/files/14531635/compare.pd.zip)

**Comment by porres** [comment]
Hey, only left inlet should take control messages, anyway, updated the first comment with the things still being discussed...

**Comment by porres** [comment]
I really feel weird about it wasting space like this 

![Screen Shot 2024-03-07 at 23 38 45](https://github.com/porres/pd-else/assets/5360142/01c990c9-56c7-4bd2-bedd-cbcd0d860a0e)

**Comment by porres** [comment]
also, should th grid be like in scope?

**Comment by porres** [comment]
range should just be regular amplitude range... not range for each dimension!

**Comment by ben-wes** [comment]
i changed the way that the display is filled in my current branch. it looks ok if you don't rotate:
![image](https://github.com/porres/pd-else/assets/1923787/0925726c-15ec-40bb-9d8f-aa6bbc32353c)

... but to me, it feels really weird when rotating then:
![image](https://github.com/porres/pd-else/assets/1923787/c2b2b740-9362-4b41-bc88-149ffbe1bd37)

... if the empty space is the problem, i could also change the behaviour and scale to the max(width, height).

**Comment by ben-wes** [comment]
@porres can you check whether the behaviour of the rotation parameters seems right to you?

currently, `rotationx` rotates around the x axis (and imho, in the wrong direction). it might not be logical - but in a way, i feel like a horizontal and vertical rotation (similar to the mouse-dragging) might be more logical?

**Comment by porres** [comment]
> i changed the way that the display is filled in my current branch. it looks ok if you don't rotate

can you send a PR so I can test?

**Comment by porres** [comment]
I don't know about rotation, is there a discrepancy?

**Comment by porres** [comment]
> I don't know about rotation, is there a discrepancy?

seems rotatex and rotatey must be swapped, right?

**Comment by porres** [comment]
so, one last thing, how do you feel about a 'range' message instead of zoom?

**Comment by ben-wes** [comment]
to me it's not obvious how to adapt this. it's also not really logical for `oscope~` ... the range setting there is useful for the single-input behavior since then you can set the min/max values that you want to display on the vertical scale. but with 2 inputs, you set min/max for both axis which to me is a bit restrictive.

but of course we can make the behavior for `scope3d~` similar and have a `range <min> <max>` setting that applies to all 3 axis. so `range -0.5 0.5` would be similar to the current `zoom 2`. feels a bit like complicating things - but we can do that change (or keep zoom as an alternative option).

what really bothers me a bit is the fact that with the non-proportional scaling and a range that applies to all axis, you can never get a proportional 3d display in the case of width != height.

**Comment by porres** [comment]
zoom seems useful to keep, range for each dimension seems overkill

**Comment by porres** [comment]
>  range for each dimension seems overkill

but you can go for it I guess, 

we'd also have to do it for [scope~]

**Comment by ben-wes** [comment]
keeping the discussion here since those PR comments will be lost in time ...

if i could redesign scope~ and scope3d~, i'd do the following:
* add a `xrange` and `yrange` to `scope~` (these names are adapted from `slider2d` and i think, they fit quite well)
* omit the `range` option in `scope3d~` completely, since it's not really predictable for anyone what it would do
* just keep `zoom` instead ... this doesn't match the behaviour of `scope~` obviously, but we shouldn't try to imitate scope~ - a 3d display is something different after all
* don't do non-proportional scaling in 3d since it creates weird skewed effects and perspectives
* rename `rotatex` to `hrotate` since it's now horizontal rotation and not a rotation around the x-axis anymore (thinking of `hsl` and `hslider` here concerning the h/v-prefix)
* rename `rotatey` to `vrotate` for the same reason
* `rotate` can then be documented accordingly (first value will change the `azimuth` and second the `elevation` ... those would be good names instead of `hrotate` and `vrotate` as well btw - but would users understand them better if they're not familiar with spherical coordinates?)
* set `nsamples` default of scope to 8. imho, it's updating way too slow with the current default and doesn't display things well with higher frequencies
* add `clip` state to `scope3d~` ... that's certainly what i'll try next now! :)

... let me know if that stuff sounds reasonable to you.

**Comment by porres** [comment]
hey, let's get back to this! There are still like 3 tasks to go, huh?

**Comment by ben-wes** [comment]
ah! I remember now. problem is: I'm still not sure I can solve the dollarsign issue since that's a requirement for pdlua. the clip is included, but performs very badly. and range is still not clear to me in a 3d context. I assume that scope~ can't be changed much for compatibility reasons. but did you check my comments above?

**Comment by porres** [comment]
> I'm still not sure I can solve the dollarsign issue since that's a requirement for pdlua.

@timothyschoen , you solved $0 expansion, right?

> did you check my comments above?

no :)

> I assume that scope~ can't be changed

we can do whatever

so, first things first, $0 expansion, I'm still landing back here and taking care of other stuff, but tomorrow and this weekend I'm busy doing live sountracks, so next week is when I'm back in for good

**Comment by ben-wes** [comment]
> solved $0 expansion, right?

just to quickly add here: what we'd need in this case is that the 'dollargs' only get expanded internally when the object is initialized, but not not when it's saved?

**Comment by porres** [comment]
So, this doesn't work

![Screen Shot 2024-05-25 at 00 37 53](https://github.com/porres/pd-else/assets/5360142/a8526f3a-e908-4273-b60c-b9d962892d84)


in compiled objects, the $0 is saved escaped and not expanded, then it expands at creation

**Comment by porres** [comment]
I remember @timothyschoen did work this out at one point, but we're not using it in [circle]

**Comment by ben-wes** [comment]
> in compiled objects, the $0 is saved escaped and not expanded, then it expands at creation

yep, that's still missing imho. and I should have mentioned this version to set it via message, too (where the message obviously needs to escape the dollar sign). I only had the creation argument in mind here.
